### ISSUE TITLE
NCS36510 build errors with ARM compiler

### ISSUE THREAD

**Issue by maclobdell** [ARMmbed/oob-mbed-os-example-mesh-minimal#7]
with the following changes to mbed_app.json

        "radio-type":{
            "help": "options are ATMEL, MCR20, NCS36510",
            "value": "NCS36510"
        },

            "target.features_add": ["NANOSTACK", "LOWPAN_HOST", "COMMON_PAL"],

mbed compile -m NCS36510 -t ARM

Results in: 

Error: L6218E: Undefined symbol mbedtls_ctr_drbg_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_random (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_seed (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_add_source (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_func (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_close_notify (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_authmode (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_ca_chain (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_handshake_timeout (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_max_version (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_min_version (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_own_cert (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_rng (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_defaults (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_get_verify_result (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_handshake_step (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_read (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_session_reset (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_set_bio (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_set_timer_cb (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_setup (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_write (referred from coap_security_handler.o).
Finished: 0 information, 0 warning and 29 error messages.
[ERROR] Error: L6218E: Undefined symbol mbedtls_ctr_drbg_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_random (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ctr_drbg_seed (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_add_source (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_func (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_entropy_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_close_notify (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_authmode (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_ca_chain (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_handshake_timeout (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_max_version (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_min_version (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_own_cert (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_conf_rng (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_defaults (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_config_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_free (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_get_verify_result (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_handshake_step (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_init (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_read (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_session_reset (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_set_bio (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_set_timer_cb (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_setup (referred from coap_security_handler.o).
Error: L6218E: Undefined symbol mbedtls_ssl_write (referred from coap_security_handler.o).
Finished: 0 information, 0 warning and 29 error messages.

GCC passes

**Comment by c1728p9** [comment]
You'll need to turn on null entropy to get this to work. The only reason GCC compiles is because it is permissive about missing functions as long as they aren't referenced.

**Comment by MarceloSalazar** [comment]
> You'll need to turn on null entropy to get this to work.

I disagree. This example is not using TLS and it should not require the user to add any TLS setting.

It is something that needs to be fixed elsewhere:
https://github.com/ARMmbed/oob-mbed-os-example-client/issues/1#issuecomment-254483281

**Comment by hasnainvirk** [comment]
mbed-os build system issue. Use the following line to get it fixed.
"macros": ["MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES","MBEDTLS_TEST_NULL_ENTROPY"],

Although we don't use mbedtls in this example, but that's what you need to do.

**Comment by kjbracey-arm** [comment]
Thread does use DTLS, so as such requires an mbed TLS build that provides the necessary functions. Which currently means having some sort of entropy source.

In theory it should be possible to link if Thread support in Nanostack was compiled away, but I believe we're falling foul of the "we're including everything in the program" build process with armlink - the unused-if-no-Thread parts of the CoAP service library are still being included in the program, and armlink is correctly reporting the missing definitions. (The C standard requires there to be exactly one definition of every object referenced in an expression in a program, and requires a diagnostic for that).

Normally this would be avoided by the CoAP service library actually being linked as a library - those library objects then wouldn't be pulled into the program in the first place if not referenced themselves. That's how it worked in Yotta, I believe.

Don't know if there's any quick-and-dirty workaround for armlink - some sort of "treat every individual object file like a library" flag. Although that in turn would probably lead to its own problems - hundreds of circular library references.... Maybe just a flag to treat it as a warning rather than error?

**Comment by kjbracey-arm** [comment]
> What you are not using , must not be compiled in.

There's no issue there. If those functions existed, there would be no diagnostic, and they would then be eliminated by the linker.

**Comment by ashok-rao** [comment]
After adding TLS macro, application works fine with ARMCC and GCC. However, still fails with IAR.
```
Compile [ 97.1%]: sleep.c
Compile [ 97.5%]: serial_api.c
Compile [ 97.9%]: sleep_api.c
Compile [ 98.3%]: spi_api.c
Compile [ 98.8%]: swversion.c
Compile [ 99.2%]: MCR20Drv.c
Compile [ 99.6%]: test_env.cpp
Compile [100.0%]: NanostackRfPhyMcr20a.cpp
[Warning] NanostackRfPhyMcr20a.cpp@131,14: [Pe550]: variable "rf_phy_channel" was set but never used
[Warning] NanostackRfPhyMcr20a.cpp@381,14: [Pe177]: function "rf_ack_wait_timer_stop" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@775,14: [Pe177]: function "rf_on" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@587,14: [Pe177]: function "rf_poll_trx_state_change" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@404,14: [Pe177]: function "rf_set_mac_address" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@682,14: [Pe177]: function "rf_cca_abort" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@392,16: [Pe177]: function "rf_read_mac_address" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@289,14: [Pe177]: function "rf_read_random" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@764,14: [Pe177]: function "rf_calibration_cb" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@1082,14: [Pe177]: function "rf_init_phy_mode" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@317,14: [Pe177]: function "rf_calibration_timer_interrupt" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@357,14: [Pe177]: function "rf_calibration_timer_start" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@328,14: [Pe177]: function "rf_cca_timer_interrupt" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@369,18: [Pe177]: function "rf_cca_timer_start" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@409,16: [Pe177]: function "rf_get_phy_mtu_size" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@891,17: [Pe177]: function "rf_tx_power_set" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@913,16: [Pe177]: function "rf_tx_power_get" was declared but never referenced
[Warning] NanostackRfPhyMcr20a.cpp@925,0: [Pe177]: function "rf_enable_antenna_diversity" was declared but never referenced
Link: oob-mbed-os-example-mesh-minimal
Error[Lp011]: section placement failed
            unable to allocate space for sections/blocks with a total estimated
                      minimum size of 0xd7ec bytes (max align 0x8) in
                      <[0x3fff4000-0x3fffffff]> (total uncommitted space
                      0xbe00).
[ERROR] Error[Lp011]: section placement failed
            unable to allocate space for sections/blocks with a total estimated
                      minimum size of 0xd7ec bytes (max align 0x8) in
                      <[0x3fff4000-0x3fffffff]> (total uncommitted space
                      0xbe00).

[mbed] ERROR: "python" returned error code 1.
[mbed] ERROR: Command "python -u C:\Ashok\mbed_CLI\OoB_round3\mesh_minimal\NCS36510\oob-mbed-os-example-mesh-minimal\mbed-os\tools\make.py -t IAR -m NCS36510 --source . --build .\BUILD\NCS36510\IAR" in "C:\Ashok\mbed_CLI\OoB_round3\mesh_minimal\NCS36510\oob-mbed-os-example-mesh-minimal"
```

**Comment by c1728p9** [comment]
Hi @ashok-rao, your most recent comment is a linker error due to running out of space and should probably be opened in a separate issue.

**Comment by hasnainvirk** [comment]
@ashok-rao I second Russ. If you look at tools/profiles/default.json, GCC is optimizing for space, and ARMCC is optimizing with level 3. May be you need a similar flag for IAR too, which would optimize for space. Or, use 6LOWPAN_HOST/THREAD_HOST mode. Routing capable devices tend to occupy more space.

**Comment by maclobdell** [comment]
I can confirm that this issue goes away when enabling Null entropy.  It would better if a useful error message would occur or avoid the problem altogether.

**Comment by c1728p9** [comment]
I created https://github.com/ARMmbed/oob-mbed-os-example-mesh-minimal/pull/10 to make missing entropy more obvious and I created https://github.com/ARMmbed/mbed-os/pull/3066 to fix the NCS36510's IAR build with this example.

**Comment by sbutcher-arm** [comment]
The fix of defining `MBEDTLS_TEST_NULL_ENTROPY ` is inappropriate and is a misuse of the feature. Please don't use this feature as a solution for anything other than debugging.

**Comment by SeppoTakalo** [comment]
Should we then just drop the support of every board without entropy?
At least for anything more complex than blinky.

**Comment by kjbracey-arm** [comment]
The chip does have a hardware TRNG - just needs the driver to be implemented.

**Comment by SeppoTakalo** [comment]
I agree with @sbutcher-arm that no examples should define the `NULL_ENTROPY` by default.

But we could link into mbed OS handbook to point out that this particular device can only be used for testing/developing:
https://docs.mbed.com/docs/mbed-os-handbook/en/5.1/advanced/tls_porting/#how-to-test-without-entropy-sources

But that should be it. We should not make it any more easier for users to make unsafe&unsecure applications. Workaround is already documented, so link into it.

**Comment by sbutcher-arm** [comment]
@andresag01 has submitted [PR #3088](https://github.com/ARMmbed/mbed-os/pull/3088) to address this issue.

The fix doesn't stop the build breaking, but should change the error to a clearer one to indicate that the entropy source is absent, and therefore the example cannot be built.
